include/tudat/astro/system_models/selfShadowing.h

```diff
@@ -0,0 +1,491 @@
+        // initialize pixelation matrix with -1 values (W in literature)
+        std::vector< std::vector< int > > pixelationMatrix( gridCoordinatesM.size(), std::vector< int >(gridCoordinatesL.size(), -1));
+        // apply PIP to shadowed plate to find its discretized surface (assign value of 0)
+        pixelationMatrix = arePointsInTriangle( allPanels_[index]->getSelfProjection( ), 
+            gridCoordinatesL, gridCoordinatesM, pixelationMatrix );
+
+        for ( int j = 0; j< static_cast< int >( sigmaMatrix_.size( ) ); j++)
+        {
+            if (sigmaMatrix_[index][j] != 0)
+            {
+                continue;
+            }
+            // resize pixelation grid to accelerate the process
+            int indexMinL = clamp(static_cast< int >( 
+                std::floor((projections_[index][j].getMinL( )-minL)/sizeL)) - 1, 0, static_cast< int >( numberL ) - 1);
+            int indexMaxL = clamp(static_cast< int >( 
+                std::ceil((projections_[index][j].getMaxL( )-minL)/sizeL)) + 1, indexMinL + 1, static_cast< int >( numberL ));
+            int indexMinM = clamp(static_cast< int >(
+                std::floor((projections_[index][j].getMinM( )-minM)/sizeM)) - 1, 0, static_cast< int >( numberM ) - 1);
+            int indexMaxM = clamp(static_cast< int >(
+                std::ceil((projections_[index][j].getMaxM( )-minM)/sizeM)) + 1, indexMinM + 1, static_cast< int >( numberM ));
+            // apply PIP to shadowing and update pixelation matrix (assign value of 1)  
+            pixelationMatrix = arePointsInTriangle(projections_[index][j], 
+                gridCoordinatesL, gridCoordinatesM, pixelationMatrix, indexMinL, indexMaxL, indexMinM, indexMaxM);
+        }
+        double fractionOnes = 0;
+        for (int p = 0; p< static_cast< int >( gridCoordinatesM.size() ); p++) 
+        {
+            for (int q = 0; q< static_cast< int >( gridCoordinatesL.size() ); q++) 
+            {
+                if (pixelationMatrix[p][q] == 1) 
+                {
+                    fractionOnes++;
+                }
+            }
+        }
+        double fractionZeros = 0;
+        for (int p = 0; p< static_cast< int >( gridCoordinatesM.size() ); p++) 
+        {
+            for (int q = 0; q< static_cast< int >( gridCoordinatesL.size() ); q++) 
+            {
+                if (pixelationMatrix[p][q] == 0) 
+                {
+                    fractionZeros++;
+                }
+            }
+        }
+        
+        fractions[i] = clamp( 1 - fractionOnes / (fractionOnes + fractionZeros ), 0.0, 1.0 );
+    }
+    return fractions;
+}
+
+class SelfShadowing
+{
+private:
+    std::vector< double > illuminatedPanelFractions_;
+public:
+    SelfShadowing( ) : illuminatedPanelFractions_( ) { };
+
+    SelfShadowing( const std::vector< std::shared_ptr< system_models::VehicleExteriorPanel > >& allPanels,

```

Comment:
Try to break this functionality up into sub-functions, to make the code easier to read/maintain